src/monodroid/jni/android-system.cc

```diff
@@ -731,89 +732,95 @@ AndroidSystem::get_gref_gc_threshold ()
+	if (v == nullptr || *v == '\0')
+		v = "";
 
-	if ((k = entry [0]) && *k &&
-			(v = entry [1]) && *v) {
-		if (islower (k [0])) {
-			add_system_property (k, v);
-		} else {
-			setenv (k, v, 1);
+	if (isupper (k [0]) || k [0] == '_') {
+		if (k [0] == '_') {
+			if (strcmp (k, "__XA_DSO_IN_APK") == 0) {
+				knownEnvVars.DSOInApk = true;
+				return;
+			}
 		}
-	}
-
-	utils.monodroid_strfreev (entry);
-}
 
-void
-AndroidSystem::setup_environment_from_file (const char *apk, int index, int apk_count, void *user_data)
-{
-	unzFile file;
-	if ((file = unzOpen (apk)) == nullptr)
+		setenv (k, v, 1);
 		return;
+	}
 
-	if (unzLocateFile (file, "environment", 0) == UNZ_OK) {
-		unz_file_info info;
+	if (k [0] == 'm') {
+		if (strcmp (k, "mono.aot") == 0) {
+			if (*v == '\0') {
+				knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_NONE;
+				return;
+			}
 
-		if (unzGetCurrentFileInfo (file, &info, nullptr, 0, nullptr, 0, nullptr, 0) == UNZ_OK &&
-				unzOpenCurrentFile (file) == UNZ_OK) {
-			char *contents = new char [info.uncompressed_size+1];
-			if (contents != NULL &&
-					unzReadCurrentFile (file, contents, info.uncompressed_size) > 0) {
+			switch (v [0]) {
+				case 'n':
+					knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_NORMAL;
+					break;
 
-				int i;
-				char *line = contents;
-				contents [info.uncompressed_size] = '\0';
+				case 'h':
+					knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_HYBRID;
+					break;
 
-				for (i = 0; i < info.uncompressed_size; ++i) {
-					if (contents [i] != '\n')
-						continue;
+				case 'f':
+					knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_FULL;
+					break;
 
-					contents [i] = '\0';
-					setup_environment_from_line (line);
-					line = &contents [i+1];
-				}
+				default:
+					knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_UNKNOWN;
+					break;
+			}
 
-				if (line < (contents + info.uncompressed_size))
-					setup_environment_from_line (line);
+			if (knownEnvVars.MonoAOT != MonoAotMode::MONO_AOT_MODE_UNKNOWN)
+				log_info (LOG_DEFAULT, "Mono AOT mode: %s", v);
+			else
+				log_warn (LOG_DEFAULT, "Unknown Mono AOT mode: %s", v);
 
-				free (contents);
-			}
+			return;
+		}
 
-			unzCloseCurrentFile (file);
+		if (strcmp (k, "mono.llvm") == 0) {
+			knownEnvVars.MonoLLVM = true;
+			return;
 		}
 	}
 
-	unzClose (file);
+	add_system_property (k, v);
 }
 
 void
-AndroidSystem::for_each_apk (JNIEnv *env, jobjectArray runtimeApks, void (AndroidSystem::*handler) (const char *apk, int index, int apk_count, void *user_data), void *user_data)
+AndroidSystem::setup_environment (JNIEnv *env, jobjectArray environmentVariables)
 {
-	int i;
-	jsize apksLength = env->GetArrayLength (runtimeApks);
-	for (i = 0; i < apksLength; ++i) {
-		jstring e       = reinterpret_cast<jstring> (env->GetObjectArrayElement (runtimeApks, i));
-		const char *apk = env->GetStringUTFChars (e, nullptr);
-
+	jsize envvarsLength = env->GetArrayLength (environmentVariables);
+	if (envvarsLength == 0)
```

Comment:
We should assert that `(envvarsLength % 2) == 0`, otherwise we'll have an exception thrown in `env->GetObjectArrayElement(e, i+1)`.